quantity/core.py

```diff
@@ -0,0 +1,250 @@
+    return v_unit.to(unit, v_value)
+
+
+def _make_op(op, helper):
+    if op.startswith("__r"):
+        op_func = getattr(operator, op.replace("__r", "__"))
+
+        def wrapped_helper(u1, u2):
+            return helper(op_func, u2, u1)
+    else:
+        op_func = getattr(operator, op)
+
+        def wrapped_helper(u1, u2):
+            return helper(op_func, u1, u2)
+
+    def __op__(self, other):
+        return self._operate(other, op, wrapped_helper)
+
+    return __op__
+
+
+def _make2arg_ops(op):
+    np_func = {
+        "__add__": np.add,
+        "__floordiv__": np.floor_divide,
+        "__matmul__": np.matmul,
+        "__mod__": np.mod,
+        "__mul__": np.multiply,
+        "__sub__": np.subtract,
+        "__truediv__": np.true_divide,
+    }[op]
+    helper = UFUNC_HELPERS[np_func]
+    fop = _make_op(op, helper)
+    rop = _make_op("__r" + op[2:], helper)
+    iop = _make_op("__i" + op[2:], helper)
+    return fop, rop, iop
+
+
+def _make_comp(comp):
+    def __comp__(self, other):
+        try:
+            other = value_in_unit(other, self.unit)
+        except Exception:
+            return NotImplemented
+        return getattr(self.value, comp)(other)
+
+    return __comp__
+
+
+def _make_deferred(attr):
+    # Use array_api_compat getter if available (size, device), since
+    # some array formats provide inconsistent implementations.
+    attr_getter = getattr(array_api_compat, attr, operator.attrgetter(attr))
+    def deferred(self):
+        return attr_getter(self.value)
+
+    return property(deferred)
+
+
+def _make_same_unit_method(attr):
+    if array_api_func := getattr(array_api_compat, attr, None):
+        def same_unit(self, *args, **kwargs):
+            return replace(self, value=array_api_func(self.value, *args, **kwargs))
+
+    else:
+        def same_unit(self, *args, **kwargs):
+            return replace(self, value=getattr(self.value, attr)(*args, **kwargs))
+
+    return same_unit
+
+
+def _make_same_unit_attribute(attr):
+    attr_getter = getattr(array_api_compat, attr, operator.attrgetter(attr))
+    def same_unit(self):
+        return replace(self, value=attr_getter(self.value))
+
+    return property(same_unit)
+
+
+def _make_defer_dimensionless(attr):
+    def defer_dimensionless(self):
+        try:
+            return getattr(self.unit.to(DIMENSIONLESS, self.value), attr)()
+        except Exception as exc:
+            raise TypeError from exc
+
+    return defer_dimensionless
+
+def _check_pow_args(exp, mod):
+        if mod is not None:
+            return NotImplemented
+
+        if not isinstance(exp, PYTHON_NUMBER):
+            try:
+                exp = exp.__complex__()
+            except Exception:
+                try:
+                    return exp.__float__()
+                except Exception:
+                    return NotImplemented
+
+        return exp.real if exp.imag == 0 else exp
+
+
+@dataclass(frozen=True, eq=False)
```

Comment:
One of the major annoyances of `dataclasses` is that they chose not to implement the `converter` argument to the `fields`. https://peps.python.org/pep-0712/. I think this was an atrocious decision that makes everyone's lives more difficult. For us, this means it's much harder to help users  when `value` is not an Array API-compatible object, like converting a `float` to a `PythonNumber[float]`-wrapped value.

Thankfully there is a well-defined way around this - the [`dataclass_transform`](https://typing.readthedocs.io/en/latest/spec/dataclasses.html#dataclass-transform) decorator. I'd be happy to contribute this in a followup PR. Essentially we produce our own `dataclass` decorator which calls python's built-in `dataclass` but provides its own `__init__`.

```
def __init__(self, *args, **kwargs):
    fs = fields(self)
    # Iterate through the fields applying the Field.metadata["converter"] then setting the attribute.
    object.__setattr__(self, f.name, value)
```

Reply:
Sounds like a plan!